gnovm/pkg/gnolang/alloc.go

```diff
@@ -104,10 +114,90 @@ func (alloc *Allocator) Allocate(size int64) {
 
 	alloc.bytes += size
 	if alloc.bytes > alloc.maxBytes {
-		panic("allocation limit exceeded")
+		if alloc.heap != nil {
+			deleted := alloc.heap.MarkAndSweep()
+			alloc.DeallocDeleted(deleted)
+
+			v, _ := mem.VirtualMemory()
+			ca := uint64(size) > v.Available
+
+			if (alloc.detonate || ca) && alloc.bytes > alloc.maxBytes {
+				panic("allocation limit exceeded")
+			}
+			alloc.detonate = alloc.bytes > alloc.maxBytes
+		}
+	}
+}
+
+func (alloc *Allocator) DeallocDeleted(objs []*GcObj) {
+	for _, obj := range objs {
+		alloc.DeallocObj(obj.tv)
+	}
+}
+
+func (alloc *Allocator) AllocateObj(tv TypedValue) {
+	switch v := tv.V.(type) {
+	case PointerValue:
+		alloc.AllocateType()
+
+		if v.TV != nil {
+			alloc.AllocateObj(*v.TV)
+		}
+	case *StructValue:
+		alloc.AllocateStruct()
+		alloc.AllocateStructFields(int64(len(v.Fields)))
+		alloc.AllocateType()
+		alloc.AllocateHeapItem()
+
+		for _, field := range v.Fields {
+			alloc.AllocateObj(field)
+		}
+	case *SliceValue:
+		alloc.AllocateSlice()
+	case *ArrayValue:
+		alloc.AllocateDataArray(int64(len(v.Data)))
+	default:
+	}
+}
+
+func (alloc *Allocator) DeallocObj(tv TypedValue) {
+	switch v := tv.V.(type) {
+	case PointerValue:
+		alloc.DeallocateType()
+		alloc.DeallocatePointer()
+	case *StructValue:
+		alloc.DeallocateStruct()
+		alloc.DeallocateStructFields(int64(len(v.Fields)))
+		alloc.DeallocateType()
+		alloc.DeallocateHeapItem()
```

Comment:
ditto.

Reply:
same as above

Reply:
in this example, the struct is not being tracked?
```go
package main

type S struct {
	num int
}

func foo() {
	s := S{num: 1}

}
func main() {
	foo()
}
```

Reply:
This is not a heap object.

Reply:
So currently, when a block is popped, the deallocation of items is only an approximation and not entirely accurate without count this in some way (sync from our previous discussion).